dbms/src/Flash/Mpp/MPPTunnel.h

```diff
@@ -301,25 +303,37 @@ class LocalTunnelSender : public TunnelSender
     {
         if (packet->hasError())
         {
-            closeLocalTunnel(true, packet->error());
+            prepareToCloseLocalTunnel(true, packet->error());
             return true;
         }
         return false;
     }
 
-    void closeLocalTunnel(bool meet_error, const String & local_err_msg)
+    // Need to tell receiver that the local tunnel will be closed and the receiver should
+    // close channels otherwise the MPPTask may hang.
+    void prepareToCloseLocalTunnel(bool meet_error, const String & local_err_msg)
     {
         bool expect = false;
         if (is_done.compare_exchange_strong(expect, true))
         {
             consumer_state.setMsg(local_err_msg);
-            local_request_handler.connectionLocalDone(meet_error, local_err_msg);
+            local_request_handler.prepareToCloseLocalConnection(meet_error, local_err_msg);
         }
     }
 
+    // It should only be called in the destructor.
```

Comment:
If it is only called in the destructor, why not just call `local_request_handler.closeLocalConnection();` in the destructor?

Reply:
> If it is only called in the destructor, why not just call `local_request_handler.closeLocalConnection();` in the destructor?

Local tunnel may push to a full queue, and may be blocked, so we need to close the channel in advande otherwise the push is blocked and `closeLocalConnection` is never called.

Reply:
Oh, what I wants to say is that since `closeLocalConnection` can only be called in the destructor, and this function itself is very simple, then why need to add `closeLocalConnection` as a new function, we can just call `local_request_handler.closeConnection();` in the destructor.